agent/consul/acl.go

```diff
@@ -25,392 +24,627 @@ const (
 	// are not allowed to be displayed.
 	redactedToken = "<hidden>"
 
-	// Maximum number of cached ACL entries.
-	aclCacheSize = 10 * 1024
+	// TODO (ACL-V2) - Is 10240 enough? In a DC with 30k agents we can only
+	//   cache 1/3 of the tokens if 1 is given to each agent
+	// identityCacheSize is the number of identities that can be cached
+	identityCacheSize = 10 * 1024
+
+	// TODO (ACL-V2) - 512 should be enough right?
+	// The max number of policies that will be cached.
+	policyCacheSize = 512
+
+	// TODO (ACL-V2) 1024 should be enough right?
+	// authorizerCacheSize is the maximum number of authorizers that may be cached
+	// The authorizer cache is access via a hash key that takes into account the
+	// policies that get compiled into the authorizer. Therefore this size is the
+	// amount of unique policy combinations which can be cached.
+	authorizerCacheSize = 1024
+
+	// aclUpgradeBatchSize controls how many tokens we look at during each round of upgrading. Individual raft logs
+	// will be further capped using the aclBatchUpsertSize. This limit just prevents us from creating a single slice
+	// with all tokens in it.
+	aclUpgradeBatchSize = 128
+
+	// aclBatchDeleteSize is the number of deletions to send in a single batch operation. 4096 should produce a batch that is <150KB
+	// in size but should be sufficiently large to handle 1 replication round in a single batch
+	aclBatchDeleteSize = 4096
+
+	// aclBatchUpsertSize is the target size in bytes we want to submit for a batch upsert request. We estimate the size at runtime
+	// due to the data being more variable in its size.
+	aclBatchUpsertSize = 256 * 1024
+
+	// DEPRECATED (ACL-Legacy-Compat) aclModeCheck* are all only for legacy usage
+	// aclModeCheckMinInterval is the minimum amount of time between checking if the
+	// agent should be using the new or legacy ACL system. All the places it is
+	// currently used will backoff as it detects that it is remaining in legacy mode.
+	// However the initial min value is kept small so that new cluster creation
+	// can enter into new ACL mode quickly.
+	aclModeCheckMinInterval = 50 * time.Millisecond
+
+	// aclModeCheckMaxInterval controls the maximum interval for how often the agent
+	// checks if it should be using the new or legacy ACL system.
+	// TODO (ACL-V2) - is this too short - potentially every 10 seconds we have to
+	// check the serf metadata for every LAN member
+	aclModeCheckMaxInterval = 30 * time.Second
```

Comment:
Well if/when we remove the legacy compatibility layer all the machinery to auto-detect whether the cluster is capable of new ACLs will go away. So my hope was that this would be sufficient for a few major releases.

I will add a follow on task to make them configurable though.

Reply:
I meant all the cache sizes maybe need to be configurable later not the aclModeCheckInterval specifically. Think you got that but just to be clear!